File: tictactoe.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

void print_board(char* state) {
    char* board = "\
┏━━━┳━━━┳━━━┓\n\
┃ x ┃ x ┃ x ┃\n\
┣━━━╋━━━╋━━━┫\n\
┃ x ┃ x ┃ x ┃\n\
┣━━━╋━━━╋━━━┫\n\
┃ x ┃ x ┃ x ┃\n\
┗━━━┻━━━┻━━━┛\n";
    do {
        if (*board == 'x') {
            putchar(*state++);
        } else {
            putchar(*board);
        }
    } while (*(++board));
}

int get_input_pos(const char current_state[9], char sign, char target, const char* question) {
    int pos;
    printf("Spelare %c, vilken ruta vill du flytta ditt %c %s ", sign, sign, question);
    fflush(stdout);
    while(!scanf("%i", &pos) || !(0 <= --pos && pos < 9) || current_state[pos] != target) {
        while ((getchar()) != '\n');
        fputs("Platsen du angav var inte giltig, var vänlig försök igen. ", stdout);
        fflush(stdout);
    }
    return pos;
}

const uint16_t winning_lines[8] = {0b111000000, 0b000111000, 0b000000111,
                                   0b100100100, 0b010010010, 0b001001001,
                                   0b100010001, 0b001010100};

uint16_t owned(const char* state, char sign) {
    uint16_t bitset = 0;
    for (int i = 0; i < 9; i++) {
        if (state[i] == sign) {
            bitset++;
        } bitset *= 2;
    }
    return bitset;
}

int main() {
    char current_state[9];
    for (int i = 0; i < 9; i++) current_state[i] = ' ';

    puts("Såhär ser din spelplan ut: ");
    print_board("123456789");
    puts("Nu börjar det, håll i er! :)");


    for (int i = 0;; i++) {
        const char sign = (i % 2 == 0) ? 'O' : 'X';

        if (i < 6) {
            int move_to = get_input_pos(current_state, sign, ' ', "till? Skriv ett tal mellan 1-9:");
            current_state[move_to] = sign;
        } else {
            int move_from = get_input_pos(current_state, sign, sign, "ifrån?");
            int move_to = get_input_pos(current_state, sign, ' ', "till?");
            current_state[move_from] = ' ';
            current_state[move_to] = sign;
        }

        print_board(current_state);

        for(int j = 0; j < 8; j++) {
            if (owned(current_state, sign) == winning_lines[i]) {
                return 0;
            }
        }
    }
}

File: tictactoe.py
import itertools


winning_lines = [
        {0, 1, 2}, {3, 4, 5}, {6, 7, 8},  # Horizontal
        {0, 3, 6}, {1, 4, 7}, {2, 5, 8},  # Vertical
        {0, 4, 8}, {2, 4, 6}              # Diagonal
    ]


def game_repr(state):
    board = """┏━━━┳━━━┳━━━┓
┃ x ┃ x ┃ x ┃
┣━━━╋━━━╋━━━┫
┃ x ┃ x ┃ x ┃
┣━━━╋━━━╋━━━┫
┃ x ┃ x ┃ x ┃
┗━━━┻━━━┻━━━┛"""
    board_pieces = board.split('x')
    acc = board_pieces[0]
    for i in range(9):
        acc += state[i] + board_pieces[i + 1]
    return acc


def get_input_pos(current_state, sign, target, question):
    while True:
        done = True
        try:
            place = int(input(f"Spelare {sign}, vilken ruta vill du flytta ditt {sign} {question} ")) - 1
            if not (0 <= place < 9 and current_state[place] == target):
                raise ValueError
        except ValueError:
            done = False
            print("platsen du angav var inte giltig, var vänlig försök igen.")
        finally:
            if done:
                return place


def main():
    current_state = [' '] * 9

    print("Såhär ser din spelplan ut: ")
    print(game_repr("123456789"))
    print("Nu börjar det, håll i er! :)")

    sign = 'O'
    for i in itertools.count():
        sign = 'X' if sign == 'O' else 'O'

        if i < 6:
            move_to = get_input_pos(current_state, sign, ' ', "till? Skriv ett tal mellan 1-9:")
        else:
            move_from = get_input_pos(current_state, sign, sign, "från?")
            move_to = get_input_pos(current_state, sign, ' ', "till?")
            current_state[move_from] = ' '

        current_state[move_to] = sign
        print(game_repr("".join(current_state)))

        owned_positions = filter(lambda pos: current_state[pos] == sign, range(9))
        if any(map(lambda line: line.issubset(owned_positions), winning_lines)):
            print(f"{sign} vann! Grattis!")
            break


main()
